### ISSUE TITLE
How to ignore defined "defaults" section of primary config.yaml while working with experiments

### ISSUE THREAD

**Issue by NinjaAtWork** [facebookresearch/hydra#1745]
Hi 
In my scenario `experiment1.yaml` has it own default section and values but hydra will sum together both defaults section of `experiment1.yaml `and primary `config.yaml` .
how to ignore primary defaults section while experimenting? 
thanks

config.yaml
```
defaults:
  - nodes.one : "Ninja"
```

experiment1.yaml
```
defaults:
  - nodes.two : "Gaiden"
```
 
result(prompt using _rich tree_)  `python main.py +experiment=experiment1`

```
CONFIG
└── NODES
      ├── one: "Ninja"
      │   
      └── two: "Gaiden"
```

desired result :
```
CONFIG
└── NODES
      └── two: "Gaiden"
```

**Comment by Jasha10** [comment]
Hi @NinjaAtWork,
I'm a little bit confused.What is the directory layout that you are using on disk?

**Comment by Jasha10** [comment]
I think there are ways to achieve what you are trying to do. It will probably involve using the override keyword (see [here](https://hydra.cc/docs/next/advanced/defaults_list#introduction)) in the defaults list of `experiment1.yaml`.

**Comment by NinjaAtWork** [comment]
Hi @Jasha10 
let me describe in this way :
the app structure(scenario) will always read `nodes `sub items and print them. 
in the primary config file i should have a `nodes.basic : "hello world"` in the `defaults`
and in experiment `defaults` i should have a `nodes.advance : "E=mc2" `
now the app will print both basic and advance sub item from nodes in experiment mode
i need to ignore the primary config yaml `defaults` section( or part of it) while doing experiments ,while the `override` will replace value, not removing config group itself
thanks

**Comment by NinjaAtWork** [comment]
@omry 
null works but:
`null` mentioned as a _Place-Holder_ for upcoming values but in my scenario using `null `option as `override `in **experiments** (not primary config) , force me consider two new rule: 
- since i am unable to add `null` as an **"OR"** option in primary config and have to use `- override config_name : null` in all experiments then any change in primary should be applied in experiments.yaml and in my app ,with many growing  experiments(used as preset for images different type) this could end up a bit simpler to manage, it was perfect if i had something new like [null-in-all-experiments-but-works-in-primary] beside [optional|override] 
- no more use for `??? `option: 
`??? `in all our `base_blahblah.yaml `which got initialize in primary config now should have value at first because null override in experiment works only in `defaults` not for override of part of same config group at remain of primary .yaml
by example:

old_config.yaml

```
defaults:
  - images@nodes.one : Ninja
nodes:
    one:
        threshold : 20
```
old images/Ninja.yaml
`threshold : ???`


new_config.yaml
```
defaults:
  - images@nodes.one : Ninja
```

new images/Ninja.yaml
`threshold : 20`

beside of these two null works as defined and please aware me if i am wrong in the way of defining my config structure.

**Comment by omry** [comment]
I don't know if I can help you further. if you have a concrete question please ask it.
Your comments are pretty hard to follow.

My suggestion is to embrace that config composition is an additive process.
The fact that you need to remove something from your config is an indication that something may be off.

**Comment by NinjaAtWork** [comment]
well, while running the app, how can i find that it started under experiment mode or not?

**Comment by Jasha10** [comment]
> well, while running the app, how can i find that it started under experiment mode or not?

Are you asking about the difference between this
```
python main.py +experiment=experiment1
```
and this
```
python main.py
```
?

My suggestion would be to inspect the `cfg` object that gets passed to your app.
You can use `cfg.keys()` to see what keys are there.

In your original example, it looks like `cfg.nodes.one` and `cfg.nodes.two` might or might not be present, depending on what command line argument you use.
You can check `hasattr(cfg.nodes, "one")` and `hasattr(cfg.nodes, "two")` to see if one or both of these is present.

**Comment by NinjaAtWork** [comment]
@Jasha10 
By far there is two way that i can pass defaults section out of primary config.
-directly from main.py code that add complexity to the code which kinda make nightmares for api parse at future.
-use null in all experiments which add difficulties mentioned above.
Could this be a feature request for experiments?
Master Omry suggest additive configs which i think has Inconsistency with the nature of the experiments

**Comment by Jasha10** [comment]
@NinjaAtWork
I think I understand the issue.
Did you consider Omry's suggestion to use the `--config-name` flag (it's documented [here](https://hydra.cc/docs/advanced/hydra-command-line-flags))?
Something like this:
```
# not running an experiment:
$ python main.py
# running an experiment:
$ python main.py --config-name=experiment_config_base +experiment=experiment1
```
You could then define two top-level configs, `config.yaml` (which would be used by default) and `experiment_config_base.yaml` (which gets used if you want to avoid the defaults list from `config.yaml`).

Another idea could be to move your `config.yaml` defaults list to a different file, e.g. `base_defaults.yaml`. Then you could add `base_defaults` at the command line if you are not experimenting.

**Comment by NinjaAtWork** [comment]
@Jasha10 
exactly my answer  `--config-name` 👍 
sorry i didn't get solution of Mr.Omry, honestly i thought it will change specific config `group` name instead of `file` name
thanks
